### ISSUE TITLE
libinih.so and libINIReader.so should be linked statically

### ISSUE THREAD

**Issue by kakra** [FeralInteractive/gamemode#222]
**Describe the bug**
When installing into a system without inih present, gamemode installs the headers and library as if they were their own package. Also, it installs pkgconfig files. This created a shadow/fake dependency. When later installing inih as a native system library, package managers will see file collisions.

Also, software may see inih pkgconfig as a dependency which is unrelated to gamemode.

Gentoo installation QA found this issue because in a multilib install, inih headers were present in the 64-bit install phase but not in the 32-bit install phase (Gentoo does not compile and install the tools and examples in 32-bit multilib just the daemon libs itself).

**To Reproduce**
1. Install gamemode without system inih present (using bundled lib)

**Expected behavior**
gamemode should not install inih libs system-wide, either it should link them statically or install them to a private lib directory. Also, the installed `gamemode_client.h` is totally unrelated to inih, so there's even no point in providing inih header and pkgconfig files.

I suggest the following files not being present in gamemode installed with bundled inih:

```
/usr/include/ini.h
/usr/include/INIReader.h
/usr/lib64/libinih.so.48
/usr/lib64/pkgconfig/INIReader.pc
/usr/lib64/pkgconfig/inih.pc
```

**System Info (please complete the following information):**
 - OS and version: Gentoo Linux
 - GameMode Version 1.5.1

**Additional context**
inih itself is unrelated to anything users or developers could do with `gamemode_client.h`, thus that dependency should be hidden from system linker view when bundled, either by linking statically or by placing in a private lib directory. I'd prefer static linking, the headers and pkgconfig have to be omitted to not collide with the package manager.

As a side note, meson should not fall back to bundled library automagically. It should always bundle it unless explicitly told otherwise. Or vice-versa and require explicit bundling if system-provided inih is missing. Usually, software bundling libs will always bundle them unless told to use the system-provided lib, so I'd prefer that variant.

**Comment by stephanlachnit** [comment]
> When installing into a system without inih present, gamemode installs the headers and library as if they were their own package. Also, it installs pkgconfig files. This created a shadow/fake dependency. When later installing inih as a native system library, package managers will see file collisions.

That's true, it is not ideal at the moment as a fallback. I would recommend the Gentoo gamemode maintainer to just pack inih, then this doesn't happen.
To solve this it would require meson changes to [upstream inih](https://github.com/benhoyt/inih). Ideally you would want an option to either install it (meaning setting `default_library=both`, installing the header and the libraries) or to build a static target (meaning setting `default_library=static` and unsetting the install option for the libraries).
Since I'm not aware that one can set options for subprojects (I might be wrong here), the default should be the static version. I don't think is easily fixed in gamemode.

**Comment by kakra** [comment]
> > When installing into a system without inih present, gamemode installs the headers and library as if they were their own package. Also, it installs pkgconfig files. This created a shadow/fake dependency. When later installing inih as a native system library, package managers will see file collisions.
> 
> That's true, it is not ideal at the moment as a fallback. I would recommend the Gentoo gamemode maintainer to just pack inih, this is the easiest fix.

Actually, I am the maintainer... ;-)

> To solve this it would require meson changes to [upstream inih](https://github.com/benhoyt/inih). Ideally you would want an option to either install it (meaning setting `default_library=both`, installing the header and the libraries) or to build a static target (meaning setting `default_library=static` and unsetting the install option for the libraries).
> Since I'm not aware that one can set options for subprojects (I might be wrong here), the default should be the static version. I don't think this is easily fixed in gamemode.
> If you want to do a PR you can /cc me, I would review it 

This requires time. As a work-around I'm removing the inih headers and pkgconfig from the build before packaging. Those should really not be installed by gamemode. But there's no such easy solution for moving (potentially) conflicting inih solib out of the way.

> > As a side note, meson should not fall back to bundled library automagically. It should always bundle it unless explicitly told otherwise.
> 
> No it should, this is exactly what it is intended for. If you really really don't want to use a fallback, you can do so by using the `--wrap-mode=nofallback` option (this is for example the default option when building a package with the debian build system, here you don't want to do that). You still have to satisfy the dependency somehow though to continue. User that install from source have no reason to use that option.
> 
> > Usually, software bundling libs will always bundle them unless told to use the system-provided lib, so I'd prefer that variant.
> 
> What? No, this is how Windows handles things, you want to use system libs as much as possible to prevent rebuilding on every new release of your dependency. Also, gamemode does bundle inih (kind of) via wrap as a fallback. If for some reason you don't want to use the system lib go with the `--wrap-mode=forcefallback` option. That this doesn't work nicely here has nothing to do with meson's behavior when having fallbacks.

It's all true what you write for the actual meson part - especially in the light of project / sub-project. It wasn't my intention to blame meson on that, it's probably handling this exactly correct. It's more about it's configuration for this project: Gamemode is to be distributed as a single package, it shouldn't contain two, with inih being some sort of shadow project.

But I took some notes from your explanation: I should look into `--wrap-mode=forcefallback` because I'm not the end-user, I'm the packager, so I have well-definied behavior in my ebuild. I still need to figure out what to do about the .so file without resorting to maintaining a second ebuild for inih.

**Comment by stephanlachnit** [comment]
> Gamemode is to be distributed as a single package, it shouldn't contain two, with inih being some sort of shadow project.

When it comes to building a package for a distro I couldn't agree more here. I just don't think there is a simple solution.

> I still need to figure out what to do about the .so file without resorting to maintaining a second ebuild for inih.

I don't know much about the Gentoo building toolchain tbh, but inih is quite simple to build, `meson builddir --prefix=/usr && ninja install -C builddir` if you do it from hand, it doesn't have any dependencies.

**Comment by Saroumane** [comment]
Hello,
as a workaround (Ubuntu 19.10) I installed "libinih-dev" and "libinih1" from official repo.
They are not seen by gamemode install script because inih.pc is missing in /usr/share/pkgconfig/
So after the 1st run of ./bootstrap.sh, I copy inih.pc provided by gamemode :  
$ sudo cp ./build/meson-private/inih.pc /usr/share/pkgconfig
Then I edit it to correct the "Version" : it has to match my Ubuntu inih package (currently 45)
Finally I restart from scratch the whole bootstrap process, inih is detected, I can install gamemode normally, and I hope I won't have packages conflicts.

**Comment by stephanlachnit** [comment]
> Do you think it's a good way to handle the problem ?

It will be fixed with #207 anyway.

**Comment by kakra** [comment]
This has been fixed for me as of #228 tho I don't know if this is due to using wrapmode force-fallback or some change of the PR.
